### PR TITLE
BUG: Avoid try/except in blocks, fix setitem bug in datetimelike EA

### PR DESCRIPTION
**Pull request by jbrockmendel**
The initial impetus for this was avoiding two try/excepts in Block methods (i.e. the diff in internals.blocks).  This uncovered the bug in DTA/TDA, which accounts for the rest of the diff.

### CODE REVIEW

File: pandas/tests/arrays/test_datetimes.py [hunk 1]
```diff
@@ -179,6 +179,22 @@ def test_setitem_clears_freq(self):
...
+    @pytest.mark.parametrize(
+        "obj",
+        [
+            pd.Timestamp.now(),
+            pd.Timestamp.now().to_datetime64(),
+            pd.Timestamp.now().to_pydatetime(),
+        ],
+    )
+    def test_setitem_objects(self, obj):
+        # make sure we accept datetime64 and datetime in addition to Timestamp
```
Comment (jreback) [inline pandas/tests/arrays/test_datetimes.py]: this is worth a release note?

File: pandas/tests/test_base.py [hunk 2]
```diff
@@ -418,7 +418,7 @@ def test_value_counts_unique_nunique_null(self, null_obj):
                     values = o._shallow_copy(v)
                 else:
                     o = o.copy()
-                    o[0:2] = iNaT
```
Comment (jreback) [inline pandas/tests/test_base.py]: what would happen to the arrange if you tried to assign iNaT? raise? coerce to object? do we have a test for that?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/arrays/test_datetimes.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimes.py]: only affects the EAa, should't be user-facing

File: pandas/tests/test_base.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/test_base.py]: coerce to object, and yes

**Review by TomAugspurger — APPROVED** [review]

**Comment by jreback** [comment]
does this need rebase? as it included the just merged set indexing patch

**Comment by jbrockmendel** [comment]
> does this need rebase?

should be OK as is

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**